File: src/lib/api/translate.ts
/**
 * Translation API Client
 * Azure Translator를 활용한 번역 API
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

// 지원 언어 타입
export type LanguageCode = 'ko' | 'en' | 'ja' | 'zh-Hans' | 'zh-Hant' | 'es' | 'fr' | 'de' | 'pt' | 'ru' | 'vi' | 'th' | 'id';

export interface TranslateRequest {
  text: string;
  to_language: LanguageCode;
  from_language?: LanguageCode;
}

export interface TranslateResponse {
  success: boolean;
  translated_text?: string;
  detected_language?: string;
  to_language?: string;
  error?: string;
}

export interface DetectResponse {
  success: boolean;
  language?: string;
  language_name?: string;
  score?: number;
  error?: string;
}

export interface LanguagesResponse {
  languages: Record<string, string>;
}

/**
 * 텍스트 번역
 */
export async function translateText(
  text: string,
  toLanguage: LanguageCode,
  fromLanguage?: LanguageCode
): Promise<TranslateResponse> {
  try {
    const response = await fetch(`${API_BASE_URL}/translate/translate`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        text,
        to_language: toLanguage,
        from_language: fromLanguage,
      }),
    });

    if (!response.ok) {
      const error = await response.json();
      return {
        success: false,
        error: error.detail || 'Translation failed',
      };
    }

    return await response.json();
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Network error',
    };
  }
}

/**
 * 언어 감지
 */
export async function detectLanguage(text: string): Promise<DetectResponse> {
  try {
    const response = await fetch(`${API_BASE_URL}/translate/detect`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ text }),
    });

    if (!response.ok) {
      const error = await response.json();
      return {
        success: false,
        error: error.detail || 'Detection failed',
      };
    }

    return await response.json();
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Network error',
    };
  }
}

/**
 * 지원 언어 목록 조회
 */
export async function getSupportedLanguages(): Promise<LanguagesResponse> {
  try {
    const response = await fetch(`${API_BASE_URL}/translate/languages`);

    if (!response.ok) {
      throw new Error('Failed to fetch languages');
    }

    return await response.json();
  } catch (error) {
    // 기본 언어 목록 반환
    return {
      languages: {
        ko: '한국어',
        en: 'English',
        ja: '日本語',
        'zh-Hans': '简体中文',
        'zh-Hant': '繁體中文',
        es: 'Español',
        fr: 'Français',
        de: 'Deutsch',
      },
    };
  }
}

// 언어 표시 이름 (프론트엔드용)
export const LANGUAGE_LABELS: Record<LanguageCode, string> = {
  ko: '한국어',
  en: 'English',
  ja: '日本語',
  'zh-Hans': '简体中文',
  'zh-Hant': '繁體中文',
  es: 'Español',
  fr: 'Français',
  de: 'Deutsch',
  pt: 'Português',
  ru: 'Русский',
  vi: 'Tiếng Việt',
  th: 'ไทย',
  id: 'Indonesia',
};

File: backend/app/routers/translate.py
"""
Translation Router
- Azure Translator API를 활용한 번역 엔드포인트
"""

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel
from typing import Optional

from ..services.translator import (
    translate_text,
    detect_language,
    get_supported_languages,
)

router = APIRouter()


class TranslateRequest(BaseModel):
    """번역 요청 모델"""
    text: str
    to_language: str  # 대상 언어 코드 (예: 'ko', 'en')
    from_language: Optional[str] = None  # 원본 언어 (None이면 자동 감지)


class TranslateResponse(BaseModel):
    """번역 응답 모델"""
    success: bool
    translated_text: Optional[str] = None
    detected_language: Optional[str] = None
    to_language: Optional[str] = None
    error: Optional[str] = None


class DetectRequest(BaseModel):
    """언어 감지 요청 모델"""
    text: str


class DetectResponse(BaseModel):
    """언어 감지 응답 모델"""
    success: bool
    language: Optional[str] = None
    language_name: Optional[str] = None
    score: Optional[float] = None
    error: Optional[str] = None


class LanguagesResponse(BaseModel):
    """지원 언어 응답 모델"""
    languages: dict


@router.post("/translate", response_model=TranslateResponse)
async def translate(request: TranslateRequest):
    """
    텍스트 번역

    - **text**: 번역할 텍스트
    - **to_language**: 대상 언어 코드 (ko, en, ja, zh-Hans 등)
    - **from_language**: 원본 언어 코드 (선택, 없으면 자동 감지)
    """
    if not request.text.strip():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Text cannot be empty"
        )

    if len(request.text) > 50000:  # Azure 제한: 50,000자
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Text exceeds maximum length (50,000 characters)"
        )

    result = await translate_text(
        text=request.text,
        to_language=request.to_language,
        from_language=request.from_language,
    )

    if not result["success"]:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=result.get("error", "Translation failed")
        )

    return TranslateResponse(**result)


@router.post("/detect", response_model=DetectResponse)
async def detect(request: DetectRequest):
    """
    텍스트 언어 감지

    - **text**: 언어를 감지할 텍스트
    """
    if not request.text.strip():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Text cannot be empty"
        )

    result = await detect_language(request.text)

    if not result["success"]:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=result.get("error", "Language detection failed")
        )

    return DetectResponse(**result)


@router.get("/languages", response_model=LanguagesResponse)
async def languages():
    """
    지원 언어 목록 조회

    번역에 사용할 수 있는 언어 코드와 이름을 반환합니다.
    """
    return LanguagesResponse(languages=get_supported_languages())
